### ISSUE TITLE
Should infer that resolved value of awaited generic Promise does not extend Promise.

### ISSUE THREAD

**Issue by ChrisHamilton91** [microsoft/TypeScript#49946]
# Bug Report

### 🔎 Search Terms

await promise generic infer type

### 🕗 Version & Regression Information

v4.8.0-dev.20220718

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.8.0-dev.20220718#code/C4TwDgpgBAcg9sACgJzgWwJYGcIB4AqAfFALxT5QQAewEAdgCZZQrrZ4CudA1nXAO51iAfih0IANwjIoALnIBuAFBKAZlwDGwDHDpRVcOAUIAKAIbIA5vPhJUmHMYCUUAN5KoUDbqxwANhAAdH5wluZWTsoAvioA9LHkABbYUFiJcBx+DFD8cMjcORjA6RzAUIlmEhh0llDAcF5mWGUWtQC0UAAGZvxmRVCtnY10fGXIEMAcyHpmdCDF1bXFZmXUtIzMrA4QSvFQAKLIqDLeaBDMqvZ1idCqZlrX0KdgGAEyvALMFBqzlDT02S27AANFAAEalKAMODnOgAcjKH341xWjyg418fik2QkZj8HCes1GcQSAywIDoGn0mm0unBFmM4WsLHs7Gcbl2pM8Bjg5l6-VakSge2kx3kAEErBwznQynBVHVwNA4eL+bQGMY4VAUqMyVgMJY6GYwQE6g0wBYzGdaDJ5YrIFA4bYgY4iHDApzPDFPfhksxcvlmPrKU90C83lAkcwMf5sVBcfjoNDzmIEH91oDWTglE0KVT1JTaXpmshFgAxOhM+QuvAlxaEFzubmGPl9FoRaIkpIpLAcDQac5YdR+PwgKAhSzMYrQOs1ShSPT8IqJMTndVQMBZ8458mU6mFnR6Sz0aQYDQVxmtatb9lN-Qt1sCiJk2AIGvOTtqGmHuprkyNjxUmAUsagrEwa0CGMsQgEwACIAEk4TQKAACsOGaAYgJAyxYKcSJAOPcRS3PSsIKgqRwK3SDzljGCEKQ7UZlXZoIGyTc2BwXC8M-Whmn-ZRPCAA)

### 💻 Code

```ts
type NotPromise<T> = T extends Promise<unknown> ? never : T;

function foo<T>(arg: NotPromise<T>) {
  console.log(arg);
}

// This should work without having to cast arg - `await arg` cannot return anything that extends Promise
// Error comes from the fact the compiler knows T can extend Promise, but doesn't know that the resolved value cannot
 async function bar<T>(arg: Promise<T>) {
   foo(await arg); // error: Argument of type 'Awaited<T>' is not assignable to parameter of type 'NotPromise<T>'.
 }

// This works since compiler knows resolved value does not extend Promise
async function stringFn(arg: Promise<string>) {
  foo(await arg);
}

// This successfully logs the string even with nested promises
async function genericFn<T>(arg: Promise<T>) {
  foo((await arg) as NotPromise<T>);
}

function test() {
  stringFn(Promise.resolve("I'm just a string"));
  genericFn(Promise.resolve(Promise.resolve("I'm in a nested promise")));
}

test();
```

### 🙁 Actual behavior

Not sure if this is a bug or just a possible improvement.

The objective is to create a function that does not accept Promises. This works fine until called inside a function with a generic Promise as a parameter. After awaiting the Promise, the compiler still thinks the value can extend Promise.

In other words, the compiler cannot infer that an awaited value does not extend Promise.

### 🙂 Expected behavior

The compiler should be able to infer that an awaited value does not extend Promise.

**Comment by RyanCavanaugh** [comment]
> The objective is to create a function that does not accept Promises

This isn't something conditional types are capable of; this code doesn't really do what it appears to do -- a call like `foo<unknown>(somePromise)` will succeed without error.

You'd really need negated types for that, see #4196

**Comment by ChrisHamilton91** [comment]
> > The objective is to create a function that does not accept Promises
> 
> This isn't something conditional types are capable of; this code doesn't really do what it appears to do -- a call like `foo<unknown>(somePromise)` will succeed without error.
> 
> You'd really need negated types for that, see #4196

It should know that Awaited<T> does not extend Promise should it not? Passing Awaited<T> to the conditional should succeed without error, just like any other type that doesn't extend promise.

**Comment by ChrisHamilton91** [comment]
> > The objective is to create a function that does not accept Promises
> 
> This isn't something conditional types are capable of; this code doesn't really do what it appears to do -- a call like `foo<unknown>(somePromise)` will succeed without error.
> 
> You'd really need negated types for that, see #4196

That just seems like a bug with `unknown` and conditional types to me. How do we know `unknown` does not extend `Promise`? That's where we should get the not assignable error, because the conditional could go either way. Instead we get this error with `Awaited<T>`. Which doesn't make sense.

Seems to me conditionals are a bit janky:
```typescript
function foo<T>(p: T extends string ? string : number){}

function bar(p: string){
  foo(p); // Argument of type 'string' is not assignable to parameter of type 'number'.
}
```

This is using `foo<unknown>` for some reason - and unknown goes to the `false` branch for some reason.

**Comment by ChrisHamilton91** [comment]
> > The objective is to create a function that does not accept Promises
> 
> This isn't something conditional types are capable of; this code doesn't really do what it appears to do -- a call like `foo<unknown>(somePromise)` will succeed without error.
> 
> You'd really need negated types for that, see #4196

After messing with it a bit more, it seems to be a limitation of conditionals, generics, and inference.

To reproduce you need 3 conditions:

1. A function with a conditional parameter, where at least one branch is generic ( `foo<T>( p: T extends any ? any: T ){}` )
2. A call to that function with no explicit type given ( `foo(a)` )
3. The argument passed must also be generic ( `bar<T>(a: T) { foo(a) }` )

You can see examples below
```typescript
// 1-4 all work fine - just showing how the compiler normally resolves conditionals when all of the above conditions are not met

function foo1<T>(p: T extends any ? any : never){}
function bar1<T>(p: T){
  foo1(p); // Infers foo1<unknown>(p: any)
}

function foo2<T>(p: T extends never ? never : any){}
function bar2<T>(p: T){
  foo2(p); // Infers foo2<unknown>(p: any)
}

function foo3<T>(p: T extends any ? T : never){}
function bar3(p: string){
  foo3(p); // Infers foo3<string>(p: string)
}

function foo4<T>(p: T extends never ? any : T){}
function bar4(p: string){
  foo4(p); // Infers foo4<string>(p: string)
}

// Everything below gives an error
// They all say: Argument of type 'T' is not assignable to parameter of type 'T extends any ? any : T'
// They all infer foo<T>(p: T extends any ? any : T) - seems the compiler can't resolve the conditional

function foo5<T>(p: T extends any ? any : T){}
function bar5<T>(p: T){
  foo5(p);
}

function foo6<T>(p: T extends any ? any : T){}
function bar6<T>(p: T){
  foo6(p);
}

function foo7<T>(p: T extends never ? T : any){}
function bar7<T>(p: T){
  foo7(p);
}

function foo8<T>(p: T extends never ? T : any){}
function bar8<T>(p: T){
  foo8(p);
}
```
